### PR TITLE
Fix/compliant writable gw

### PR DESCRIPTION
**Pull request by cryptix**
This is follow up to #1886 ~~(and might have some commits from it until it is merged)~~ rebased.

@ion1 made a fair point [here](https://github.com/ipfs/go-ipfs/pull/1886#issuecomment-150664078) about POST being a better choice for the `ipfs object patch` functionality of the writable gateway. I whipped up some tests for these cases.

My main open question: How do we want to specify the name to overwrite in the `POST /ipfs/QmFoo/...` case? I don't like to involve the `Content-Disposition: …; filename="…"` header. Can't we just use the last part of the request path as that? Also feels like a pain to get right for users.

### CODE REVIEW

**Comment by ion1** [comment]
> My main open question: How do we want to specify the name to overwrite in the `POST /ipfs/QmFoo/...` case? I don't like to involve the `Content-Disposition: …; filename="…"` header. Can't we just use the last part of the request path as that? Also feels like a pain to get right for users.

There is a bit of a mismatch between the assumptions of the HTTP methods and our assumption of content-addressed and immutable paths. However, [the RFC text on `POST`](https://tools.ietf.org/html/rfc7231#section-4.3.3) seems much more flexible about the exact semantics than the text on `PUT` (which says “A successful PUT of a given representation would suggest that a subsequent GET on that *same* target resource will result in an equivalent representation being sent in a 200 (OK) response”).

The usual assumption about `POST /resource` is that the `/resource` already exists and the `POST` will create a subresource under it. If `/ipfs/QmBar/baz/file` does not exist, it would be surprising if there was a resource at that path that responds to `POST`. Normally one would `POST` to the existing resource `/ipfs/QmBar/baz` and specify the filename of the subresource to be created using a means that is not the path.

On the other hand, perhaps we could read the RFC liberally and interpret “The POST method requests that the target resource process the representation enclosed in the request according to the resource's own specific semantics” with the assumption that given an existing `/ipfs/QmBar`, as far as `POST` is concerned, there is a resource at all possible paths under it such that `POST /ipfs/QmBar/baz/file` patches `QmBar` and responds with `201 Created`, `Location: /ipfs/QmQuux/baz/file`.

This would certainly be much nicer for everyone than having to handle a special `Content-Disposition` header in the payload.

**Comment by cryptix** [comment]
Note to self: keep an eye on https://github.com/ipfs/infrastructure/issues/105 for relevant parts.

**Comment by jbenet** [comment]
Was this waiting on input from me?

**Comment by cryptix** [comment]
@jbenet no, sorry. I'll try to get this done this week.

**Comment by cryptix** [comment]
rebased, tests still broken.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: eb204f693d4381a19869f52726ecf98fa842261a
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help. Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

**Comment by cryptix** [comment]
Tests should be fixed now. RFCR.

**Comment by cryptix** [comment]
Forgot about the sharness tests - can't run them on my bsd office box right now, fill fix them later at home

File: core/corehttp/gateway_handler.go [hunk 1]
```diff
@@ -55,16 +55,20 @@ func (i *gatewayHandler) newDagFromReader(r io.Reader) (*dag.Node, error) {
...
 	if i.config.Writable {
 		switch r.Method {
 		case "POST":
-			i.postHandler(w, r)
+			i.postHandler(ctx, w, r)
 			return
 		case "PUT":
-			i.putHandler(w, r)
+			// TODO(cryptix): where are the docs?
+			http.Error(w, "writableGateway: PUT method not meaningful on IPFS - use POST and see the docs", http.StatusMethodNotAllowed)
```
Comment (lgierth) [inline core/corehttp/gateway_handler.go]: Can just remove the PUT-case and let it fall through to the method error in lines 81 ff.

File: core/corehttp/gateway_handler.go [hunk 2]
```diff
@@ -264,38 +269,30 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
...
-		internalWebError(w, err)
-		return
-	}
-
-	k, err := i.node.DAG.Add(nd)
-	if err != nil {
-		internalWebError(w, err)
-		return
+func (i *gatewayHandler) postHandler(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+	// TODO(cryptix): can we move this into ServeHTTP, so we can do it once for all methods?
```
Comment (lgierth) [inline core/corehttp/gateway_handler.go]: Yes can move this to ServeHTTP

**Comment by lgierth** [comment]
From ipfs/infrastructure#105

> > @jbenet: need a way to force --pin=false for all ops on the gateway.
>
> How would you like to set this? Config option oder flag on 'ipfs daemon' ?

Just unconditionally have the gateway not pin -- we can add an option for implicit pinning later (I don't think we need it).

File: core/corehttp/gateway_handler.go (hunk 2 above)
Comment (cryptix) [inline core/corehttp/gateway_handler.go]: Okay, I want to do this together with tests in another PR. I think this is big enough as it is.

File: core/corehttp/gateway_handler.go (hunk 1 above)
Comment (cryptix) [inline core/corehttp/gateway_handler.go]: m) thanks! :)

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 105aeb738dd889debf303c4ee4221c73e55600c9
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help. Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

**Comment by cryptix** [comment]
> Just unconditionally have the gateway not pin.

Thanks for that insight and the review!

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: ede0559470ce2d3ee0a21076efde5765e72b7c61
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help. Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

**Comment by jbenet** [comment]
> Just unconditionally have the gateway not pin -- we can add an option for implicit pinning later (I don't think we need it).

ok sgtm

**Comment by jbenet** [comment]
could other people CR this too?

**Comment by cryptix** [comment]
rebased and now apparently broken.. not sure what changed under this.

**Comment by jbenet** [comment]
:( --

**Comment by whyrusleeping** [comment]
just looks like your code is returning 405's for a bunch of things...

**Comment by cryptix** [comment]
I know - it passed the last time I worked on it  before the rebase. Need to take a log of the source subtree and see what was introduce.

**Comment by whyrusleeping** [comment]
@cryptix do you have time to take another look at this?

**Comment by cryptix** [comment]
Actually, Yup. I fear the rebase, though. I might need to nag somebody but will try.

**Comment by cryptix** [comment]
@whyrusleeping I tried but I have no idea what is going on with the code changes. Hope somebody else  can pick this up.
